pkg/binary/envoy/controlplane/istio_test.go

```diff
@@ -38,39 +37,57 @@ func TestMain(m *testing.M) {
+//
+// Technically what happens is bootstrap results in a "xds-grpc" cluster pointed at Pilot's gRPC address. During
+// bootstrap, Envoy invokes "/envoy.service.discovery.v3.AggregatedDiscoveryService/StreamAggregatedResources", which
+// reads configuration in pilot (from testdata/configs.yaml). This test passes when that configuration merges with the
+// bootstrap configuration generated by Istio.
 func TestConnectsToMockPilotAsAGateway(t *testing.T) {
-	pilotGrpcAddr, stopPilot := requireMockPilot(t, "testdata")
+	// only includes configs.yaml, which includes egress rules for googleapis
+	pilotServer, stopPilot := requireMockPilot(t, "testdata")
 	defer stopPilot()
 
+	pilotGrpc := pilotServer.GRPCListener.Addr().String()
 	runtime, err := envoy.NewRuntime(
 		func(r *envoy.Runtime) {
 			r.Config.Mode = envoy.ParseMode("loadbalancer")
-			r.Config.XDSAddress = pilotGrpcAddr
-			r.Config.IPAddresses = []string{"1.1.1.1"}
+			// Choosing a random port doesn't help because envoy_bootstrap.json hard-codes 15020 15021 and 15090
+			r.Config.AdminPort = 15000
+			r.Config.XDSAddress = pilotGrpc
+			r.Config.IPAddresses = []string{"127.0.0.1"} // prevent calling controlplane.retrieveIPs() on CI hosts
 		},
 		Istio,                                // this integrates and ends up reading pilotGrpcAddr
 		debug.EnableEnvoyAdminDataCollection, // this allows us to read clusters.txt
 	)
 	require.NoError(t, err, "error creating envoy runtime")
 	defer os.RemoveAll(runtime.DebugStore())
+	time.Sleep(5 * time.Second)
 
 	envoytest.RequireRunTerminate(t, runtime, envoytest.RunKillOptions{
-		// Envoy won't achieve ready status because endpoints in testdata/configs.yaml are unreachable
-		ExpectedStatus: binary.StatusInitializing,
-		// We sleep to allow the configuration to populate, so that we can read it back.
-		SleepBeforeTerminate: 5 * time.Second,
-		// Assertions below inspect files in the debug store
-		RetainDebugStore: true,
+		DisableAutoAdminPort: true,
+		RetainDebugStore:     true, // Assertions below inspect files in the debug store
 	})
 
-	// Verify configuration from istio testdata/configs.yaml ended up in envoy
-	gotClusters, err := os.ReadFile(filepath.Join(runtime.DebugStore(), "clusters.txt"))
-	require.NoError(t, err, "error getting envoy clusters")
-	require.Contains(t, string(gotClusters), "istio-ingressgateway.istio-system.svc.cluster.local::1.1.1.1:8443")
+	serverInfo, err := os.ReadFile(filepath.Join(runtime.DebugStore(), "server_info.json"))
+	require.NoError(t, err, "error getting server_info.json")
+	clusters, err := os.ReadFile(filepath.Join(runtime.DebugStore(), "clusters.txt"))
+	require.NoError(t, err, "error getting clusters.txt")
+
+	// First, ensure r.Config.XDSAddress ended up written to config as the discoveryAddress
+	require.Contains(t, string(serverInfo), fmt.Sprintf(`"discoveryAddress": "%s"`, pilotGrpc))
+
+	// Next, ensure a connection discoveryAddress was made (Envoy -> Pilot's XDS)
```

Comment:
While I can't commit to this now, you can probably see how this test could now be refactored into a unit test and an e2e test, where only the latter needs to run istio. We can mock the XDS service here in other words and doing so will massively clean up the dependency tree.